### PR TITLE
Introduce distinct ObjectMappers for RestEasy Reactive / RestEasy Reactive Client

### LINKED ISSUE #23979: Distinct ObjectMappers for RestEasy Reactive / RestEasy Reactive Client
### Description

Quarkus allows a program to use specify a custom Jackson ObjectMapper by using CDI producer method or customize the ObjectMapper with ObjectMapperCustomizer instances.

I need to have different behaviors for RestEasy Reactive Client use and RestEasy Reactive use. Specifically, I want the service implementation to be set with .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, true) and the client to have .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)

### Implementation ideas

Perhaps annotate the CDI producer method and the ObjectMapperCustomizer(s) with with a new qualifier:

```java
package io.quarkus.jackson.runtime;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import javax.inject.Qualifier;
import javax.ws.rs.RuntimeType;

@Qualifier
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.METHOD, ElementType.FIELD, ElementType.PARAMETER, ElementType.TYPE})
public @interface ObjectMapperUse {
  RuntimeType value();
}
```

How much backwards compatibility is required?  Is there a fallback to unqualified producer if qualified producer is unavailable?

### PR DESCRIPTION
**Pull request by chonton**
Closes #23979

Provide three ObjectMappers: unqualified, client, and server that can be independently produced, customized, and injected.  This allows different behaviors for the different uses.

### CODE REVIEW

**Comment by geoand** [comment]
Thanks for this.

To be honest, I would like to limit the scope of the change.

What I had in mind was to have these separate `ObjectMapper`s only for RESTEasy Reactive and also (and probably more importantly), only leverage the client and server `ObjectMapper` objects **if** they have been created. Otherwise, RESTEasy Reactive should use the default `ObjectMapper`.
The reason is that `ObjectMapper` is very very heavy (both in terms of heap space used and of time taken to create it), so the common use case where users don't need per-case instances should not suffer any performance penalty just to support the less common case.

**Comment by gsmet** [comment]
Yes, we need to be extremely careful about that.

**Comment by kdubb** [comment]
What do you guys think about using `JsonMapper` instead of `ObjectMapper` during this change? Or maybe in another PR? As of `2.10`, `JsonMapper` appears to be the preferred system.

**Comment by gsmet** [comment]
Another thing we need to be careful about is backwards compatibility. You might have people relying on the existing behavior so we need to be careful.

**Comment by chonton** [comment]
I've create #24019 to track using `JsonMapper`

**Comment by chonton** [comment]
Limited the scope of the change per @geoand 's comments

**Comment by geoand** [comment]
Thanks, I'll take a closer look soon.

**Comment by chonton** [comment]
@geoand, Rebased and addressed  concerns.  I had to move annotations to `quarkus-resteasy-reactive-jackson-common` module rather than `quarkus-resteasy-reactive-jackson` module.

Are the qualifier annotations in the package that you expect?

File: extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/runtime/ClientObjectMapper.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import javax.enterprise.util.AnnotationLiteral;
+import javax.inject.Qualifier;
+
+/**
+ * Qualifier Annotation that specializes production, customization, and injection of `client` ObjectMapper
```
Comment (geoand) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/runtime/ClientObjectMapper.java:12]: I think we should also mention what the purpose of this `client` ObjectMapper is - i.e. that it is only used for the reactive rest client

File: extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/runtime/ServerObjectMapper.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import javax.enterprise.util.AnnotationLiteral;
+import javax.inject.Qualifier;
+
+/**
+ * Qualifier Annotation that specializes production, customization, and injection of `server` ObjectMapper
```
Comment (geoand) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/runtime/ServerObjectMapper.java:12]: Same here

File: docs/src/main/asciidoc/rest-json.adoc [hunk 3]
```diff
@@ -245,6 +245,149 @@ public class CustomObjectMapper {
 }
 ----
 
+There is one unqualified `ObjectMapper` instances provided by the quarkus-jackson extension.  You may produce two additional
+`ObjectMapper` instances qualified by the `ClientObjectMapper` and `ServerObjectMapper` qualifier annotations.  If provided, the
```
Comment (geoand) [inline docs/src/main/asciidoc/rest-json.adoc:249]: `ClientObjectMapper` -> `@ClientObjectMapper`, `ServerObjectMapper` -> `@ServerObjectMapper`

**Review by geoand — COMMENTED** [review]
I added a final round of comments.

We'll also need the commits squashed.

Thanks!

File: extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/runtime/ClientObjectMapper.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package io.quarkus.resteasy.reactive.jackson.common.runtime;
```
Comment (geoand) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/runtime/ClientObjectMapper.java:1]: Actually, it's probably best if this is moved into the client module.
Also, we don't want to include `runtime` in the package.

File: extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/runtime/ServerObjectMapper.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package io.quarkus.resteasy.reactive.jackson.common.runtime;
```
Comment (geoand) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/runtime/ServerObjectMapper.java:1]: Same comment as the client

File: extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/runtime/ClientObjectMapper.java (hunk 4 above)
Comment (chonton) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/runtime/ClientObjectMapper.java]: I have moved annotations to `io.quarkus.resteasy.reactive.jackson.common` package.  Annotations remain in `quarkus-resteasy-reactive-jackson-common` module.

Moving annotations to `quarkus-resteasy-reactive-jackson` module causes build to fail with following exception:

Deployment artifact io.quarkus:quarkus-rest-client-reactive-jackson-deployment::jar:999-SNAPSHOT is missing the following dependencies from its configuration: io.quarkus:quarkus-resteasy-reactive-jackson-deployment::jar, io.quarkus:quarkus-resteasy-reactive-deployment::jar, io.quarkus:quarkus-vertx-http-deployment::jar -> [Help 1]

Fixing that causes BasicTest in `integration-tests/rest-client-reactive` module to fail.  This seems way too complicated to fix.

File: docs/src/main/asciidoc/rest-json.adoc [hunk 6]
```diff
@@ -245,6 +245,156 @@ public class CustomObjectMapper {
 }
 ----
 
+There is one unqualified `ObjectMapper` instance provided by the quarkus-jackson extension.  Your code may produce two additional
+`ObjectMapper` instances. The client `ObjectMapper` instance is qualified by the `@ClientObjectMapper` annotation. The server
+`ObjectMapper` instance is qualified by the `@ServerObjectMapper` annotation.
```
[the remark this answers is not in the document]
Comment (chonton) [reply docs/src/main/asciidoc/rest-json.adoc:250]: I am happy to change to `@RestClientObjectMapper` and `@RestServerObjectMapper` if that is consensus.

[the remark this answers is not in the document]
Comment (geoand) [reply docs/src/main/asciidoc/rest-json.adoc:250]: I like the idea

File: extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/runtime/ClientObjectMapper.java (hunk 1 above)
Comment (chonton) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/runtime/ClientObjectMapper.java]: done

File: extensions/jackson/runtime/src/main/java/io/quarkus/jackson/runtime/ClientObjectMapper.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import javax.enterprise.util.AnnotationLiteral;
+import javax.inject.Qualifier;
+
+/**
+ * Qualifier Annotation that specializes production, customization, and injection of `client` ObjectMapper
+ */
+@Qualifier
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.METHOD, ElementType.FIELD, ElementType.PARAMETER, ElementType.TYPE })
+public @interface ClientObjectMapper {
```
[the remark this answers is not in the document]
Comment (chonton) [reply extensions/jackson/runtime/src/main/java/io/quarkus/jackson/runtime/ClientObjectMapper.java]: moved annotations to io.quarkus.resteasy.reactive.jackson.common

File: docs/src/main/asciidoc/rest-json.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (chonton) [reply docs/src/main/asciidoc/rest-json.adoc:250]: Alternatively, a single annotation `@ObjectMapperTarget` with `ObjectMapperAudience` value.  `ObjectMapperAudience` is an extensible enum with current values of `REACTIVE_REST_CLIENT` and `REACTIVE_REST_SERVER`.

[the remark this answers is not in the document]
Comment (chonton) [reply docs/src/main/asciidoc/rest-json.adoc]: Reverted to `@RestClientObjectMapper` and `@RestServerObjectMapper`
Rebased and squashed.
Any other issues to address?

**Comment by geoand** [comment]
LGTM, thanks a lot!

@gsmet any more comments on your side?

**Comment by chonton** [comment]
@geoand @gsmet Ready to merge?

**Comment by geoand** [comment]
Let's wait for Guillaume's comments

File: extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/RestClientObjectMapper.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import javax.enterprise.util.AnnotationLiteral;
+import javax.inject.Qualifier;
+
+/**
+ * Qualifier Annotation that specializes production, customization, and injection of `client` ObjectMapper. If an ObjectMapper
+ * is produced with this qualifier, it is used by the quarkus-jaxrs-client-reactive extension for serialization and
+ * deserialization. Otherwise, quarkus-resteasy-reactive uses the unqualified `ObjectMapper` instance.
```
Comment (gsmet) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/RestClientObjectMapper.java:14]:
```suggestion
 * deserialization. Otherwise, quarkus-rest-client-reactive uses the unqualified `ObjectMapper` instance.
```

File: docs/src/main/asciidoc/rest-json.adoc [hunk 9]
```diff
@@ -245,6 +245,151 @@ public class CustomObjectMapper {
 }
 ----
 
+There is one unqualified `ObjectMapper` instance provided by the quarkus-jackson extension.  Your code may produce two additional
+`ObjectMapper` instances. The client `ObjectMapper` instance is qualified by the `@RestClientObjectMapper` annotation. The server
+`ObjectMapper` instance is qualified by the `@RestServerObjectMapper` annotation.
+
+If the client `ObjectMapper` instance is available, it is used by the quarkus-jaxrs-client-reactive extension for serialization
```
Comment (gsmet) [inline docs/src/main/asciidoc/rest-json.adoc:252]:
```suggestion
If the REST client `ObjectMapper` instance is available, it is used by the quarkus-rest-client-reactive extension for serialization
```

File: docs/src/main/asciidoc/rest-json.adoc [hunk 10]
```diff
@@ -245,6 +245,151 @@ public class CustomObjectMapper {
...
+and deserialization.  If the client instance is not provided, then quarkus-resteasy-reactive uses the unqualified `ObjectMapper`
+instance.
+
+If the server `ObjectMapper` is available, it is used by the quarkus-resteasy-reactive extension for serialization and
+deserialization.  If the server instance is not provided, then quarkus-resteasy-reactive uses the unqualified `ObjectMapper`
```
Comment (gsmet) [inline docs/src/main/asciidoc/rest-json.adoc:257]:
```suggestion
deserialization.  If the REST server instance is not provided, then quarkus-resteasy-reactive uses the unqualified `ObjectMapper`
```

File: docs/src/main/asciidoc/rest-json.adoc [hunk 11]
```diff
@@ -245,6 +245,151 @@ public class CustomObjectMapper {
...
+    private final ObjectMapper serializer;
+
+    @Inject
+    public UseFallback(ObjectMapper mapper,
+            @RestClientObjectMapper Instance<ObjectMapper> clientInstance) {
+        serializer = clientInstance.isUnsatisfied() ? mapper : clientInstance.get();
+    }
+
+}
+----
```
Comment (gsmet) [inline docs/src/main/asciidoc/rest-json.adoc:282]: I'm not sure we really want to document this. These mappers are supposed to be used for a very specific purposes by the RESTEasy Reactive extension so I'm not sure I would document anything except the customization of them.

In any case, if we want to keep that (I recommend we don't), I would put it at the end, not at the start.

File: extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/RestServerObjectMapper.java [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import javax.enterprise.util.AnnotationLiteral;
+import javax.inject.Qualifier;
+
+/**
+ * Qualifier Annotation that specializes production, customization, and injection of `server` ObjectMapper. If an ObjectMapper
```
Comment (gsmet) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/RestServerObjectMapper.java:12]:
```suggestion
 * Qualifier Annotation that specializes production, customization, and injection of the REST server ObjectMapper. If an ObjectMapper
```

File: docs/src/main/asciidoc/rest-json.adoc [hunk 13]
```diff
@@ -245,6 +245,151 @@ public class CustomObjectMapper {
...
+
+If you want to customize the unqualified, client, and server instances differently, you can use the `@RestClientObjectMapper` and
+`@RestServerObjectMapper` annotations on your `ObjectMapperCustomizer`.  `ObjectMapperCustomizer` instances without any
+qualifier will customize all three `ObjectMapper`s. `ObjectMapperCustomizer` annotated with `@RestClientObjectMapper` will customize
+only the client instance.  `ObjectMapperCustomizer` annotated with `@RestServerObjectMapper` will customize only the server instance.
```
Comment (gsmet) [inline docs/src/main/asciidoc/rest-json.adoc:287]: Hmmmm. With the way it's documented, I would have expected the client and server instances to be created automatically if there are customizers and AFAICS from the code, it's not the case?

I.e. I would have expected the third part of the doc where you have to produce the `ObjectMapper` yourselves to not be necessary as it's quite brittle and you could definitely get it wrong doing things by yourself.
So basically, I would have produced synthetic beans for these object mappers if the annotations were found on customizers as IIRC, it's not possible to return null if we want to produce a `@Singleton`.

@geoand does it make sense or am I missing something?

**Review by gsmet — CHANGES_REQUESTED** [review]
I added some minor suggestions but I have also one major concern we need to sort out. See inline.

File: docs/src/main/asciidoc/rest-json.adoc (hunk 13 above)
Comment (chonton) [reply docs/src/main/asciidoc/rest-json.adoc:287]: Prior comment from some reviewer asked to remove creation of client and server instances.  My quarkus mojo may not be up to the task of creating the beans as part of build.

File: docs/src/main/asciidoc/rest-json.adoc (hunk 11 above)
Comment (chonton) [reply docs/src/main/asciidoc/rest-json.adoc:282]: Makes sense to remove the fallback in this example.  When would someone wish to inject the alternate ObjectMappers into non-quarkus code?  Possibly/Probably an edge case.

File: docs/src/main/asciidoc/rest-json.adoc (hunk 13 above)
Comment (gsmet) [reply docs/src/main/asciidoc/rest-json.adoc:287]: So we don't want to create the ``ObjectMapper``s if they are not needed. My question was: could we create them only if they are required. I think we could do that with synthetic beans but I want to check with @geoand that this makes sense.

Comment (gsmet) [reply docs/src/main/asciidoc/rest-json.adoc:287]: Hum. And I'm not really sure this works because IIRC the Client writer is only initialized if the server part is not around: https://github.com/quarkusio/quarkus/blob/main/extensions/resteasy-reactive/rest-client-reactive-jackson/deployment/src/main/java/io/quarkus/rest/client/reactive/jackson/deployment/RestClientReactiveJacksonProcessor.java#L42 .

Also the `JacksonBasicMessageBodyReader` that you haven't changed and should probably be changed is shared between the two.

So my guess is that this needs more thoughts. But I might be missing something as I'm not that familiar with RESTEasy Reactive.

Comment (geoand) [reply docs/src/main/asciidoc/rest-json.adoc:287]: @gsmet so basically you are asking if we can create the "RestServerObjectMapper" if there are customizers for it?

Comment (gsmet) [reply docs/src/main/asciidoc/rest-json.adoc:287]: @geoand so that was my initial question yes. As otherwise it's very cumbersome (especially, since you have to apply the customizers yourself).

But... my other question is: I'm pretty sure the current proposal won't work because in some cases we share the writer between client and server and if I'm not mistaken the reader is always shared.
I'm wondering if we could use `@ConstrainedTo(RuntimeType.SERVER)` / `@ConstrainedTo(RuntimeType.CLIENT)` but I'm not sure it's implemented in RESTEasy Reactive.
If we can, we would make the reader/writer for client/server totally distinct and this could work.

Comment (geoand) [reply docs/src/main/asciidoc/rest-json.adoc:287]: > As otherwise it's very cumbersome (especially, since you have to apply the customizers yourself).

I think it's acceptable price to pay, at least initially until we see if people actually use this feature.

> I'm pretty sure the current proposal won't work because in some cases we share the writer between client and server and if I'm not mistaken the reader is always shared.

What reader and writter are you talking about?

Comment (geoand) [reply docs/src/main/asciidoc/rest-json.adoc:287]: Good point, I had totally forgot about that! Indeed we would need to always have both and limit them to client and server

**Comment by geoand** [comment]
I recently came across another use case where this would be very useful, I think we should reevaluate it, WDYT @gsmet?
